versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/ReferenceLogicNonTxImpl.java

```diff
@@ -0,0 +1,607 @@
+ * non-transactional databases to consistently delete a reference, where non-transactional databases
+ * additionally run logic to re-run a previously failed process to create or delete a reference.
+ *
+ * <h3>Create reference / Logical approach</h3>
+ *
+ * <p>The <em>logical</em> approach to create a reference is as follows:
+ *
+ * <ol>
+ *   <li>Create a commit in {@link InternalRefs#REF_REFS} with a {@link Action#ADD} for the
+ *       reference using a {@link RefObj} payload, which contains the reference information used
+ *       during creation. Implementations must only commit, if the reference is not present, to
+ *       allow the resume/recovery process described below.
+ *   <li>After the reference has been added to {@link InternalRefs#REF_REFS}, call {@link
+ *       Persist#addReference(Reference)}, which creates the atomically updatable entry.
+ * </ol>
+ *
+ * <p>Note: Updates to a named reference are <em>not</em> tracked in {@link InternalRefs#REF_REFS}.
+ *
+ * <h3>Delete reference / Logical approach</h3>
+ *
+ * <ol>
+ *   <li>The {@link Reference} is marked as {@link Reference#deleted()} using {@link
+ *       Persist#markReferenceAsDeleted(Reference)}.
+ *   <li>Then the reference deletion is tracked by creating a commit in {@link
+ *       InternalRefs#REF_REFS} using a {@link Action#REMOVE} operation. Implementations must only
+ *       commit, if the reference is present, to allow the resume/recovery process described below.
+ *   <li>Finally, the {@link Reference} is purged (row physically deleted) via {@link
+ *       Persist#purgeReference(Reference)}.
+ * </ol>
+ *
+ * <h3>Fetching references</h3>
+ *
+ * Accessing references by their full name is performed via {@link Persist#findReference(String)}
+ * (or {@link Persist#findReferences(String[])}) and follows the resume/recovery process described
+ * below.
+ *
+ * <h3>Listing references</h3>
+ *
+ * Listing/querying references is performed via the tip of {@link InternalRefs#REF_REFS} and then
+ * {@link Persist#findReferences(String[])} chunks of references to inquire their tips/HEADs.
+ *
+ * <h3>Non-transactional resume/recovery</h3>
+ *
+ * Transactional databases can safely use the native consistent update mechanisms provided by the
+ * database. But non-transactional databases need some logic, which works as follows whenever a
+ * reference is accessed:
+ *
+ * <ul>
+ *   <li>{@link Persist#findReference(String)} (or {@link Persist#findReferences(String[])}) is used
+ *       to fetch a reference by name.
+ *   <li>If the reference is found and {@link Reference#deleted()} is {@code false}, the reference
+ *       has been found and can be returned.
+ *   <li>If the the {@link Reference#deleted()} flag is {@code true}, it is possible that a previous
+ *       delete-reference operation failed in the middle. The implementation must resume the
+ *       reference-deletion process described above.
```

Comment:
I'd like to avoid that in the ref-logic-impl, because it would also be tricky for "tools" (what could that do/be?) w/ the reference-row and references-index.